wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/RuntimeMessageAdapter.kt

```diff
@@ -0,0 +1,204 @@
+   */
+  val jsonAlternateNames: List<String?> = fieldBindingsArray.map {
+    when {
+      it.jsonName != it.declaredName -> it.declaredName
+      it.jsonName != it.name -> it.name
+      else -> null
+    }
+  }
+
+  /** When writing each field as JSON this is the name to use. */
+  val FieldOrOneOfBinding<*, *>.jsonName: String
+    get() = if (wireFieldJsonName.isEmpty()) declaredName else wireFieldJsonName
+
+  fun newBuilder(): B = binding.newBuilder()
+
+  override fun encodedSize(value: M): Int {
+    val cachedSerializedSize = binding.getCachedSerializedSize(value)
+    if (cachedSerializedSize != 0) return cachedSerializedSize
+
+    var size = 0
+    for (field in fields.values) {
+      val fieldValue = field[value] ?: continue
+      size += field.adapter().encodedSizeWithTag(field.tag, fieldValue)
+    }
+    size += binding.unknownFields(value).size
+
+    binding.setCachedSerializedSize(value, size)
+    return size
+  }
+
+  override fun encode(writer: ProtoWriter, value: M) {
+    for (field in fields.values) {
+      val binding = field[value] ?: continue
+      field.adapter().encodeWithTag(writer, field.tag, binding)
+    }
+    writer.writeBytes(binding.unknownFields(value))
+  }
+
+  override fun redact(value: M): M {
+    val builder = binding.newBuilder()
+    for (field in fields.values) {
+      if (field.redacted && field.label == WireField.Label.REQUIRED) {
+        throw UnsupportedOperationException(
+            "Field '${field.name}' in $type is required and cannot be redacted."
+        )
+      }
+      val isMessage = field.isMessage
+      if (field.redacted || isMessage && !field.label.isRepeated) {
+        val builderValue = field.getFromBuilder(builder)
+        if (builderValue != null) {
+          val redactedValue = field.adapter().redact(builderValue)
+          field.set(builder, redactedValue)
+        }
+      } else if (isMessage && field.label.isRepeated) {
+        @Suppress("UNCHECKED_CAST")
+        val values = field.getFromBuilder(builder) as List<Any>
+
+        @Suppress("UNCHECKED_CAST")
+        val adapter = field.singleAdapter() as ProtoAdapter<Any>
+        field.set(builder, values.redactElements(adapter))
+      }
+    }
+    binding.clearUnknownFields(builder)
+    return binding.build(builder)
+  }
+
+  override fun equals(other: Any?): Boolean {
+    return other is RuntimeMessageAdapter<*, *> && other.messageType == messageType
+  }
+
+  override fun hashCode(): Int = messageType.hashCode()
+
+  override fun toString(value: M): String = buildString {
+    append(messageType.simpleName)
+    append('{')
+    var first = true
+    for (field in fields.values) {
+      val binding = field[value] ?: continue
+      if (!first) append(", ")
+      first = false
+      append(field.name)
+      append('=')
+      append(if (field.redacted) REDACTED else binding)
+    }
+    append('}')
+  }
```

Comment:
Note that I changed this method because it was missing the `}`, and because `StringBuilder.replace()` doesn’t exist on multiplatform.